rmf_fleet_adapter/src/rmf_fleet_adapter/agv/FleetUpdateHandle.cpp

```diff
@@ -557,63 +492,124 @@ void FleetUpdateHandle::Implementation::bid_notice_cb(
     }
     ++index;
   }
+
   bid_proposal_pub->publish(bid_proposal);
   RCLCPP_INFO(
     node->get_logger(),
-    "Submitted BidProposal to accommodate task [%s] with new cost [%f]",
-    id.c_str(), cost);
+    "Submitted BidProposal to accommodate task [%s] by robot [%s] with new cost [%f]",
+    id.c_str(), bid_proposal.robot_name.c_str(), cost);
 
   // Store assignments in internal map
   bid_notice_assignments.insert({id, assignments});
 
 }
 
+//==============================================================================
 void FleetUpdateHandle::Implementation::dispatch_request_cb(
   const DispatchRequest::SharedPtr msg)
 {
   if (msg->fleet_name != name)
     return;
 
   const std::string id = msg->task_profile.task_id;
-  const auto task_it = bid_notice_assignments.find(id);
 
-  if (task_it == bid_notice_assignments.end())
-    return;
+  if (msg->method == DispatchRequest::ADD)
+  {
+    const auto task_it = bid_notice_assignments.find(id);
+    if (task_it == bid_notice_assignments.end())
+      return;
 
-  RCLCPP_INFO(
-    node->get_logger(),
-    "Bid for task_id:[%s] awarded to fleet [%s]",
-    id.c_str(), name.c_str());
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Bid for task_id:[%s] awarded to fleet [%s]",
+      id.c_str(), name.c_str());
 
-  // We currently only support adding tasks
-  if (msg->method != DispatchRequest::ADD)
-    return;
-  
-  const auto& assignments = task_it->second;
-  
-  if (assignments.size() != task_managers.size())
-  {
-    RCLCPP_ERROR(
+    auto& assignments = task_it->second;
+    
+    if (assignments.size() != task_managers.size())
+    {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "The number of available robots does not match that in the assignments "
+        "for task_id:[%s]. This request will be ignored.", id.c_str());
+
+      return;
+    }
+
+    // Here we make sure none of the tasks in the assignments has already begun
+    // execution. If so, we replan assignments until a valid set is obtained 
+    // and only then update the task manager queues
+    bool valid_assignments = false;
+    while (!valid_assignments)
+    {
+      valid_assignments = is_valid_assignments(assignments);
+      if (!valid_assignments)
+      {
+        const auto request_it = generated_requests.find(id);
+        if (request_it == generated_requests.end())
+          return;
+        const auto replan_results = allocate_tasks(request_it->second);
+        if (!replan_results)
+        {
+          RCLCPP_WARN(
+            node->get_logger(),
+            "Unable to replan assignments for task_id:[%s]", id.c_str());
+          return;
+        }
+        assignments = replan_results.value();
+        // We re-check the assignments as a task could have started while replanning
+        valid_assignments = is_valid_assignments(assignments);
```

Comment:
This whole function is being called by the ROS2 executor, which means it's running on the main rxcpp worker. There isn't any risk that the assignments would be invalidated during the time it takes to replan, because no new tasks will be started while the planning is happening.

On that note, it might not be good that the replanning is blocking the main thread. If tests have been going okay, then we can leave this as-is for now, but a better non-blocking solution would be to run the replanning on a separate worker and then deliver the result back to the main worker. Let's at least leave a note here to improve this so that it doesn't block.

Reply:
Thank you for explaining this. I've removed the "re-check" for the assignments and added notes explaining the above. https://github.com/osrf/rmf_core/pull/265/commits/347e4bff842986b6ba7dc0d653bc7743f53b57b2

So far in my testing, I have not encountered any delays due to the blocking nature of this function. Will continue to keep an eye out for this and address the issue in the future.